providers/flagd/pkg/service/in_process/grpc_sync.go

```diff
@@ -207,12 +172,32 @@ func (g *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
 		}
 
 		// Attempt to create sync stream
-		if err := g.performSyncCycle(ctx, dataSync); err != nil {
+		err := g.performSyncCycle(ctx, dataSync)
+		if err != nil {
 			if ctx.Err() != nil {
 				g.Logger.Info("sync cycle failed due to context cancellation")
 				return ctx.Err()
 			}
 
+			// check for non-retryable errors during initialization, if found return with FATAL
+			if !g.IsReady() {
+				st, ok := status.FromError(err)
+				if ok {
+					if _, found := nonRetryableCodes[st.Code()]; found {
+						errStr := fmt.Sprintf("first sync cycle failed with non-retryable status: %v, "+
+							"returning provider fatal.", st.Code().String())
+						g.Logger.Error(errStr)
+						return &of.ProviderInitError{
+							ErrorCode: of.ProviderFatalCode,
+							Message:   errStr,
+						}
+					}
+				}
+			}
+
+			// Backoff before retrying
+			time.Sleep(time.Duration(g.RetryBackOffMaxMs) * time.Millisecond)
```

Comment:
I am not a big fan of our blocking sleeps, as they clearly have some disadvantages, should we maybe stick to a timer for this kind of logic? like 
```golang
select {
case <-time.After(time.Duration(g.RetryBackOffMaxMs) * time.Millisecond):
    // ... code here ...
case <-ctx.Done():
    return // Allows cancellation
}
```

Reply:
There's disadvantages of sleep, but I think right now this is better than nothing, because nothing == a tight loop in some cases that is a serious bug in some situations.